client/src/actions/todos.js

```diff
@@ -0,0 +1,62 @@
+export const FETCH_TODOS_BEGIN = 'FETCH_TASKS_BEGIN';
+export const FETCH_TODOS_SUCCESS = 'FETCH_TODOS_SUCCESS';
+export const FETCH_TODOS_FAILURE = 'FETCH_TODOS_FAILURE';
+
+export const fetchTodosBegin = () => {
+  return { type: FETCH_TODOS_BEGIN };
+};
+
+export const fetchTodosSuccess = (todos) => {
+  return { type: FETCH_TODOS_SUCCESS, payload: todos };
+};
+
+export const fetchTodosFailure = (error) => {
+  return { type: FETCH_TODOS_FAILURE, payload: error };
+};
+
+export const addTodo = (todo) => {
+  return (dispatch, getState) => {
+    dispatch(fetchTodosBegin());
+    try {
+      const todos = getState().todos.items;
+      const addedTodos = [...todos, todo];
+      dispatch(fetchTodosSuccess(addedTodos));
+    } catch (error) {
+      dispatch(fetchTodosFailure(error));
+    }
+  };
+};
+
+export const deleteTodo = (id) => {
+  return (dispatch, getState) => {
+    dispatch(fetchTodosBegin());
+    try {
+      const todos = getState().todos.items;
+      const updatedTodos = todos.filter((todo) => todo.id !== id);
+      dispatch(fetchTodosSuccess(updatedTodos));
+    } catch (error) {
+      dispatch(fetchTodosFailure(error));
+    }
+  };
+};
+
+export const toggleTodo = (id) => {
+  return (dispatch, getState) => {
+    dispatch(fetchTodosBegin());
+    try {
+      const todos = getState().todos.items;
+      const updatedTodos = todos.map((todo) => {
+        if (todo.id === id) {
+          return {
+            ...todo,
+            completed: !todo.completed,
+          };
+        }
+        return todo;
+      });
+      dispatch(fetchTodosSuccess(updatedTodos));
```

Comment:
define this functions on reducer maintain a single state

Reply:
This is redux thunk function it isn't necessary now but when dealing with asynchronous API it is very important so I used it for now too.

Reply:
Redux thunk has been defined on initializing the stores and redux actions so on single components we can take over from redux state